tests/random_string_number_tests.cpp

```diff
@@ -0,0 +1,157 @@
+  if (rand.next_bool()) {
+    buffer[pos++] = '-';
+  }
+  size_t number_of_digits = size_t(rand.next_digit_count());
+  std::uniform_int_distribution<int> decimal_generator(1,int(number_of_digits));
+  size_t location_of_decimal_separator = size_t(decimal_generator(rand.generator));
+  for (size_t i = 0; i < number_of_digits; i++) {
+    if (i == location_of_decimal_separator) {
+      buffer[pos++] = '.';
+    } 
+    if (( i == 0) && (location_of_decimal_separator != 1)) {
+      buffer[pos++] = char(rand.next_nonzero_digit() + '0');
+    } else {
+      buffer[pos++] = char(rand.next_digit() + '0');
+    }
+  }
+  if (rand.next_bool()) {
+    if (rand.next_bool()) {
+      buffer[pos++] = 'e';
+    } else {
+      buffer[pos++] = 'E';
+    }
+    if (rand.next_bool()) {
+      buffer[pos++] = '-';
+    } else {
+      if (rand.next_bool()) {
+        buffer[pos++] = '+';
+      }
+    }
+    number_of_digits = rand.next_exp_count();
+    size_t i = 0;
+    if(number_of_digits > 0) {
+        buffer[pos++] = char(rand.next_nonzero_digit() + '0');
+        i++;
+    }
+    for (; i < number_of_digits; i++) {
+      buffer[pos++] = char(rand.next_digit() + '0');
+    }
+  }
+  buffer[pos] = '\0'; // null termination
+  return pos;
+}
+
+
+#ifndef TEST_FLOATS
+// We do not recognize the system, so we do not verify our results.
+bool check_float(double , const char *) {
+    return true;
+}
+#else
+bool check_float(double result, const char *buf) {
+  char *endptr;
+#ifdef _WIN32
+  static _locale_t c_locale = _create_locale(LC_ALL, "C");
+  double expected = _strtod_l((const char *)buf, &endptr, c_locale);
+#else
+  static locale_t c_locale = newlocale(LC_ALL_MASK, "C", NULL);
+  double expected = strtod_l((const char *)buf, &endptr, c_locale);
+#endif    
+  if (endptr == (const char *)buf) {
+    fprintf(stderr,
+            "parsed %f from %.32s whereas strtod refuses to parse a float, ",
+            result, buf);
+    return false;
+  }
+  if (expected != result) {
+    fprintf(stderr, "parsed %.128e from \n", result);
+    fprintf(stderr, "       %.32s whereas strtod gives\n", buf);
+    fprintf(stderr, "       %.128e,", expected);
+    return false;
+  }
+  return true;
+}
+#endif
+
+
+/**
+ * We generate random strings and we try to parse them,
+ * and we verify that we get the same answer.
+ */
+bool tester(int seed, size_t volume) {
+  char buffer[1024]; // large buffer (can't overflow)
+  simdjson::dom::parser parser;
+  RandomEngine rand(seed);
+  double result;    
+  for (size_t i = 0; i < volume; i++) {
+    if((i%100000) == 0) { std::cout << "."; std::cout.flush(); }
+    size_t length = build_random_string(rand, buffer);
+    auto error = parser.parse(buffer, length).get(result);
+    // When we parse a (finite) number, it better match strtod.
+    if ((!error) && (!check_float(result, buffer))) { return false; } 
+  }
+  return true;
+}
+
+int main() {
+  // We test 1,000,000 random strings.
+  if (tester(1234344, 1000000)) {
```

Comment:
It would be a nice project to just test every single one of them, probably across multiple machines or over a long time period--then we'd know this will never give us a random failure. But I'm good with this, it's good extra coverage and I don't think there's a huge chance of such a thing.

Reply:
It occurs to me that there are an infinite number of potential strings, even if there is a finite number of strings we'd probably need to test (the ones on and adjacent to the border between floats). Again, this is just a comment and seriously don't do it right now :)

Reply:
I am puzzled about these comments. Yes, there are definitively infinitely many strings. I test a million of them. It obviously does not establish that there is no bug. I do this to give us *confidence* that there is no bug.

It is super hard to prove that there is no possible bug. There might be. I just try to make it as unlikely as possible.

Even just checking all 64-bit floats is impossible. At least impossible given my ressources. But even if you took all 64-bit values and serialized them and checked that you got back to original, you have to contend with the fact that there an infinite number of ways to serialize a given 64-bit value.

Isn't it just life as a programmer?

Reply:
Your test is good :) I'm mostly musing. I know Rust has claimed to have done a comprehensive test of float parsing with at least one of each possible float; a quick search didn't turn it up though, I'll keep looking.

Reply:
Suppose you spent 100 cycles on a 4 GHz machine for each one of the possible 2**64 binary64 value, it would take you 15,000 years. But, of course, you need more than one hundred cycles per float...  Still, 15,000 processors is within reach at places like Microsoft, Google and the NSA. But that would still give you only one test per float... and you would have kept busy 15,000 processors for a year. At that scale, just ensuring that the system does what you think it does is going to be a massive engineer achievement.

At the other hand of the spectrum, 1,000,000 is not much but it is easy enough to crank this up and let it run for days.

Reply:
I will change the code so that one can run it as long as one wants (but the 1,000,000 is a good default).

Reply:
Done. So the -m 32132312 would run 32132312 tests.

Reply:
Finally done.